crates/svm-cli/tests/spawn_app_tests.rs

```diff
@@ -0,0 +1,216 @@
+                .iter()
+                .take(4)
+                .collect::<Vec<_>>()
+        )
+    );
+    assert_eq!(
+        &caps[5],
+        format!("{:?}", {
+            case.ctor_args
+                .iter()
+                .map(|v| wasm_value::parse_str(v).unwrap())
+                .collect::<Vec<_>>()
+        })
+    );
+}
+
+#[test]
+fn encode_invalid_outputpath() {
+    let output_path = "";
+    let version = "0";
+    let template_addr = "00aa00aa00aa00aa00aa00aa00aa00aa00aa00aa";
+    let ctor_index = "0";
+
+    let input = vec![
+        "myprog",
+        "encode",
+        "spawn_app",
+        output_path,
+        version,
+        template_addr,
+        ctor_index,
+    ];
+
+    let matches = cli::new_app().get_matches_from(input);
+    let res = cli::process(matches);
+    assert!(res.is_err());
+    assert!(res
+        .err()
+        .unwrap()
+        .to_string()
+        .starts_with(&format!("failed to create file at {}", output_path)));
+}
+
+#[test]
+fn encode_invalid_template_addr() {
+    let output_path = "";
+    let version = "0";
+    let template_addr = "00aa00aa00aa00aa00aa";
+    let ctor_index = "0";
+
+    let input = vec![
+        "myprog",
+        "encode",
+        "spawn_app",
+        output_path,
+        version,
+        template_addr,
+        ctor_index,
+    ];
+
+    let matches = cli::new_app().get_matches_from(input);
+    let res = cli::process(matches);
+    assert!(res.is_err());
+    assert_eq!(
+        res.err().unwrap().to_string(),
+        format!(
+            "invalid template address length: found {}, expected: {}",
+            hex::decode(template_addr).unwrap().len(),
+            Address::len()
+        )
+    );
+}
+
+#[test]
+fn encode_invalid_hex() {
+    let invalid_hex_str = "00a";
+    assert!(hex::decode(invalid_hex_str).is_err());
+
+    // Invalid `template_addr`.
+    let output_path = "";
+    let version = "0";
+    let template_addr = invalid_hex_str;
+    let ctor_index = "0";
+
+    let input = vec![
+        "myprog",
+        "encode",
+        "spawn_app",
+        output_path,
+        version,
+        template_addr,
+        ctor_index,
+    ];
+
+    let matches = cli::new_app().get_matches_from(input);
+    let res = cli::process(matches);
+    assert!(res.is_err());
+    assert!(res
+        .err()
+        .unwrap()
+        .to_string()
+        .starts_with("failed to decode hex string"));
+
+    // Invalid `ctor_buf`.
+    let output_path = "";
+    let version = "0";
+    let template_addr = "00aa00aa00aa00aa00aa00aa00aa00aa00aa00aa";
+    let ctor_index = "0";
+    let ctor_buf = invalid_hex_str;
+
+    let input = vec![
+        "myprog",
+        "encode",
+        "spawn_app",
+        output_path,
+        version,
+        template_addr,
+        ctor_index,
+        ctor_buf,
+    ];
+
+    let matches = cli::new_app().get_matches_from(input);
+    let res = cli::process(matches);
+    assert!(res.is_err());
+    assert!(res
```

Comment:
please try to assert against a `struct/enum` that represents the error:
something like:

```rust
assert_eq!(Err(DecodeError::InvalidDigit)), res);
```

Reply:
The returned err type is currently `String`.
Do you mean I should create a custom struct/enum err type, or return an err of the `hex::FromHexError` type?
I didn't use either of them because the err returned from `cli::process` need to support other types as well (e.g., `io::Error`), so I thought to use `String` as the baseline format.

Reply:
Actually, I might be able to return different err types by using a trait type, but AFAIK i'm unable to prepend additional string information to the err.

Reply:
It's not that critical, but since you are the owner of the API, you can have an ad-hoc
error type (`struct / enum`) to make tests more explicit and less brittle.

Reply:
I'm just wandering what's the best practice for this situation.

Reply:
I personally prefer having an explicit `Error` type over generic `Box<dyn Error>`

Reply:
Ok, i'll fix this in a follow-up PR, because the scope of this is lager than `SpawnApp`.